dubbo-configcenter/dubbo-configcenter-zookeeper/src/main/java/org/apache/dubbo/configcenter/support/zookeeper/ZookeeperDynamicConfiguration.java

```diff
@@ -98,57 +94,47 @@ protected void initWith(URL url) {
+     * @return
+     */
     @Override
-    protected String getTargetConfig(String key, String group, long timeout) {
-        // when group is not null, we are getting startup configs from Config Center
-        // for example, group=dubbo, key=dubbo.properties
-        if (StringUtils.isNotEmpty(group)) {
-            key = group + "/" + key;
-        }
-        // when group is null, we are fetching governance rules.
-        // for example, key=org.apache.dubbo.DemoService.configurators
-        else {
-            int i = key.lastIndexOf(".");
-            key = key.substring(0, i) + "/" + key.substring(i + 1);
+    public Object getInternalProperty(String key) {
+        ChildData childData = treeCache.getCurrentData(key);
+        if (childData != null) {
+            return new String(childData.getData(), StandardCharsets.UTF_8);
         }
-
-        return (String) getInternalProperty(rootPath + "/" + key);
+        return null;
     }
 
     /**
      * For service governance, multi group is not supported by this implementation. So group is not used at present.
-     *
-     * @param key
-     * @param group
-     * @param cacheListener
-     * @param listener
      */
     @Override
-    protected void addConfigurationListener(String key, String group, CacheListener cacheListener, ConfigurationListener listener) {
+    public void addListener(String key, String group, ConfigurationListener listener) {
         cacheListener.addListener(key, listener);
     }
 
     @Override
-    protected void removeConfigurationListener(String key, String group, CacheListener cacheListener, ConfigurationListener configurationListener) {
-        cacheListener.removeListener(key, configurationListener);
+    public void removeListener(String key, String group, ConfigurationListener listener) {
+        cacheListener.removeListener(key, listener);
     }
 
     @Override
-    protected CacheListener createTargetListener(String key, String group) {
-        return cacheListener;
-    }
-
-    /**
-     * @param key e.g., {service}.configurators, {service}.tagrouters, {group}.dubbo.properties
-     * @return
-     */
-    @Override
-    protected Object getInternalProperty(String key) {
-        ChildData childData = treeCache.getCurrentData(key);
-        if (childData != null) {
-            return new String(childData.getData(), StandardCharsets.UTF_8);
+    public String getConfig(String key, String group, long timeout) throws IllegalStateException {
+        // when group is not null, we are getting startup configs from Config Center
+        // for example, group=dubbo, key=dubbo.properties
+        if (StringUtils.isNotEmpty(group)) {
+            key = group + "/" + key;
         }
-        return null;
+        // when group is null, we are fetching governance rules.
+        // for example, key=org.apache.dubbo.DemoService.configurators
+        else {
```

Comment:
is there any possibility of key having null?

Reply:
In reality, key passed to here will never be null.